### ISSUE TITLE
stable-6.0 broken by #7095

### ISSUE THREAD

**Issue by imranh2** [ceph/ceph-ansible#7139]
**Bug Report**

What happened:
* Checkout ceph-ansible stable-6.0
* Configure it do do auto osd selection
* Run site.yaml
* Errors

```
fatal: [atl-ceph-node2]: FAILED! => 
  msg: |-
    The conditional check 'ansible_facts['mounts'] | selectattr('device', 'equalto', device) | length == 0' failed. The error was: Unexpected templating type error occurred on ({% if ansible_facts['mounts'] | selectattr('device', 'equalto', device) | length == 0 %} True {% else %} False {% endif %}): object of type 'generator' has no len()
  
    The error appears to be in '/ceph-ansible/roles/ceph-facts/tasks/devices.yml': line 77, column 3, but may
    be elsewhere in the file depending on the exact syntax problem.
  
    The offending line appears to be:
  
  
    - name: set_fact devices generate device list when osd_auto_discovery
      ^ here
```

What you expected to happen:
* Checkout ceph-ansible stable-6.0
* Configure it do do auto osd selection
* Run site.yaml
* Working ceph cluster

How to reproduce it (minimal and precise):

```
containerized_deployment: false
ceph_origin: repository
ceph_repository: community
ceph_stable_release: pacific
osd_objectstore: bluestore


osd_auto_discovery: true
dmcrypt: True
osds_per_device: 4
```

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 20.04.3
* Kernel (e.g. `uname -a`): `Linux atl-ceph-node4 5.4.0-107-generic #121-Ubuntu SMP Thu Mar 24 16:04:27 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
* Docker version if applicable (e.g. `docker version`):
* Ansible version (e.g. `ansible-playbook --version`): 2.9.27
* ceph-ansible version (e.g. `git head or tag or stable branch`): stable-6.0
* Ceph version (e.g. `ceph -v`): n/a

Fix (for people stumbling on this):
* Use ceph-ansible `v6.0.25.5` tag

Investigation:
Backport PR - https://github.com/ceph/ceph-ansible/pull/7095
Actual PR - https://github.com/ceph/ceph-ansible/pull/7052

**Comment by imranh2** [comment]
@guits Sorry the Ansible run is long gone now and I've got a working ceph cluster.

#7134 appears to be the exact issue. Perhaps I can give you `ansible_facts['mounts']` from the node I quoted above to aid with your debugging?

**Comment by guits** [comment]
@imranh2 as you can see in https://github.com/ceph/ceph-ansible/pull/7134#issuecomment-1096190088 i'm just not sure how to reproduce the issue, that's why I'd like to take a look at a full log, i must be missing a detail maybe I would find out in a that log

**Comment by guits** [comment]
@clwluvw FYI

**Comment by imranh2** [comment]
@guits I can repro it. It's an Ansible version problem.

```
# cat repro-playbook.yml 
- hosts: all
  become: true
  gather_facts: true
  tasks:
    - debug:
        msg: "{{ ansible_facts['mounts'] }}"

    - name: set_fact devices generate device list when osd_auto_discovery
      vars:
        device: "{{ item.key | regex_replace('^', '/dev/') }}"
      set_fact:
        devices: "{{ devices | default([]) | union([device]) }}"
      with_dict: "{{ ansible_facts['devices'] }}"
      when:
        - ansible_facts['mounts'] | selectattr('device', 'equalto', device) | length == 0

    - debug:
        msg: "{{ devices }}"
```

Works on ansible `2.10.8` fails on `2.9.27`.

Why am I using 2.9? Because https://github.com/ceph/ceph-ansible/blob/stable-6.0/requirements.txt#L2

**Comment by imranh2** [comment]
Can confirm that the fix suggested in https://github.com/ceph/ceph-ansible/pull/7134 works

**Comment by imranh2** [comment]
Task:

```
    - debug:
        msg: "{{ ansible_facts['mounts'] | selectattr('device', 'equalto', '/dev/nvme1n1') }}"
```

2.9:

```
TASK [debug] ***********************************************************************************************************************
ok: [atl-ceph-node1] => {
    "msg": "<generator object select_or_reject at 0x7f22d82d9040>"
}
ok: [atl-ceph-node2] => {
    "msg": "<generator object select_or_reject at 0x7f22d82d9040>"
}
ok: [atl-ceph-node3] => {
    "msg": "<generator object select_or_reject at 0x7f22d83be200>"
}
ok: [atl-ceph-node4] => {
    "msg": "<generator object select_or_reject at 0x7f22d83c8200>"
}
```

2.10:

```
TASK [debug] ***********************************************************************************************************************
ok: [atl-ceph-node1] => {
    "msg": []
}
ok: [atl-ceph-node2] => {
    "msg": []
}
ok: [atl-ceph-node3] => {
    "msg": []
}
ok: [atl-ceph-node4] => {
    "msg": []
}
```

**Comment by guits** [comment]
@imranh2 thanks!!

**Comment by guits** [comment]
should be fixed when #7144 gets merged
